### PR TITLE
Word RNNLM

### PR DESCRIPTION
**Pull request by takaaki-hori**
Implemented word-based RNNLM modules, `MultiLevelLM` and `LookAheadLM`.
Currently, they can be used only for decoding.
LookAheadLM is available with `run.sh --use-wordlm true` in `egs/wsj/asr1`. In this case, a word RNNLM will be trained instead of character RNNLM, and used for decoding.
At present, only WSJ example has this option `--use-wordlm`.

### CODE REVIEW

File: egs/wsj/asr1/run.sh [hunk 1]
```diff
@@ -173,19 +175,38 @@ fi
...
+else
+    lm_batchsize=2048
+    lmexpdir=exp/train_rnnlm_2layer_bs${lm_batchsize}
+    lmdict=$dict
+fi
 mkdir -p ${lmexpdir}
 if [ ${stage} -le 3 ]; then
     echo "stage 3: LM Preparation"
     lmdatadir=data/local/lm_train
     mkdir -p ${lmdatadir}
```
[the remark this answers is not in the document]
Comment (takaaki-hori) [reply egs/wsj/asr1/run.sh]: Thanks, that's right. I will fix it soon.

File: src/asr/asr_chainer.py [hunk 2]
```diff
@@ -506,6 +507,31 @@ def recog(args):
...
+                s, i = ln.split()
+                word_dict[s] = int(i)
+            if '<eos>' not in word_dict:
+                word_dict['<eos>'] = len(word_dict)
+        else:
+            word_dict = None
+
+        char_dict = {x: i for i, x in enumerate(train_args.char_list)}
+
+        word_rnnlm = lm_chainer.ClassifierWithState(lm_chainer.RNNLM(len(word_dict), 650))
```
[the remark this answers is not in the document]
Comment (takaaki-hori) [reply src/asr/asr_chainer.py]: I think it is better that such hyper parameters are read from a file, e.g. lm.conf, or the model file itself to keep asr_recog.py simple. We need more discussions about how to handle LM files and their hyper parameters. So I'd like to consider this issue some other day.

File: src/lm/extlm_chainer.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+    '''
+    # node [dict(subword_id -> node), word_id, word_set[start-1, end]]
+    root = [{}, -1, None]
+    for w, wid in word_dict.items():
+        if wid > 0 and wid != word_unk:  # skip <blank> and <unk>
+            if True in [c not in subword_dict for c in w]:  # skip unknown subword
+                continue
+            succ = root[0]  # get successors from root node
+            for i, c in enumerate(w):
```
[the remark this answers is not in the document]
Comment (takaaki-hori) [reply src/lm/extlm_chainer.py]: I have checked the code generates identical tree structure with python2 and python3.

**Comment by sw005320** [comment]
Thanks!
Do you get better results in the WSJ task?
Can you also update `RESULTS` and `README.md` if there are some improvement?

**Comment by sw005320** [comment]
rebase is OK. I'll do final check.

**Pull request merged**